Sources/Sentry/Profiling/SentryProfiler.mm

```diff
@@ -0,0 +1,190 @@
+}
+
+- (instancetype)init {
+    if (self = [super init]) {
+        _debugImageProvider = [[SentryDebugImageProvider alloc] init];
+    }
+    return self;
+}
+
+- (void)start {
+    // Disable profiling when running with TSAN because it produces a TSAN false
+    // positive, similar to the situation described here:
+    // https://github.com/envoyproxy/envoy/issues/2561
+    #if defined(__has_feature)
+    #if __has_feature(thread_sanitizer)
+    return;
+    #endif
+    #endif
+    @synchronized(self) {
+        if (_profiler != nullptr) {
+            _profiler->stopSampling();
+        }
+        _profile = [NSMutableDictionary<NSString *, id> dictionary];
+        const auto sampledProfile = [NSMutableDictionary<NSString *, id> dictionary];
+        const auto samples = [NSMutableArray<NSDictionary<NSString *, id> *> array];
+        const auto threadMetadata = [NSMutableDictionary<NSString *, NSDictionary *> dictionary];
+        sampledProfile[@"samples"] = samples;
+        sampledProfile[@"thread_metadata"] = threadMetadata;
+        _profile[@"sampled_profile"] = sampledProfile;
+        _referenceUptimeNs = getReferenceTimestamp();
+        
+        __weak const auto weakSelf = self;
+        _profiler = std::make_shared<SamplingProfiler>([weakSelf, sampledProfile, threadMetadata, samples](auto &backtrace) {
+            const auto strongSelf = weakSelf;
+            if (strongSelf == nil) {
+                return;
+            }
+            assert(backtrace.uptimeNs >= strongSelf->_referenceUptimeNs);
+            const auto threadID = [@(backtrace.threadMetadata.threadID) stringValue];
+            if (threadMetadata[threadID] == nil) {
+                const auto metadata = [NSMutableDictionary<NSString *, id> dictionary];
+                if (!backtrace.threadMetadata.name.empty()) {
+                    metadata[@"name"] = [NSString stringWithUTF8String:backtrace.threadMetadata.name.c_str()];
+                }
+                metadata[@"priority"] = @(backtrace.threadMetadata.priority);
+                threadMetadata[threadID] = metadata;
+            }
+#if defined(DEBUG)
+            const auto symbols = backtrace_symbols(
+                          reinterpret_cast<void *const *>(backtrace.addresses.data()), static_cast<int>(backtrace.addresses.size()));
+#endif
+            const auto frames = [NSMutableArray<NSDictionary<NSString *, id> *> new];
+            for (std::vector<uintptr_t>::size_type i = 0; i < backtrace.addresses.size(); i++) {
+                const auto frame = [NSMutableDictionary<NSString *, id> dictionary];
+                frame[@"instruction_addr"] = sentry_formatHexAddress(@(backtrace.addresses[i]));
+#if defined(DEBUG)
+                frame[@"function"] = [NSString stringWithUTF8String:symbols[i]];
+#endif
+                [frames addObject:frame];
+            }
+
+            const auto sample = [NSMutableDictionary<NSString *, id> dictionary];
```

Comment:
Most definitely, this is a first pass at a very naive implementation of this solution. The Foundation data structures are inefficient for this purpose especially since this data structure will keep reallocating and growing as we append samples to it, it's next on my list to work on a more optimized solution.